File: subversion/bindings/javahl/native/RevisionRange.cpp
/**
 * @copyright
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 * @endcopyright
 *
 * @file RevisionRange.cpp
 * @brief Implementation of the class RevisionRange.
 */

#include <apr_pools.h>
#include "svn_client.h"
#include "svn_types.h"

#include "JNIUtil.h"
#include "RevisionRange.h"
#include "Revision.h"
#include "Pool.h"

RevisionRange::RevisionRange(jobject jrevisionRange)
{
    m_range = jrevisionRange;
}

RevisionRange::~RevisionRange()
{
    // m_range is assume to be managed externally, and thus is not
    // explicitly destroyed.
}

namespace {
void get_range_info(jobject jrange,
                    svn_opt_revision_t* range_start,
                    svn_opt_revision_t* range_end,
                    svn_boolean_t* range_inheritable)
{
  JNIEnv *env = JNIUtil::getEnv();

  jclass clazz = env->FindClass(JAVAHL_CLASS("/types/RevisionRange"));
  if (JNIUtil::isExceptionThrown())
    return;

  if (range_start)
    {

      static jmethodID fmid = 0;
      if (fmid == 0)
        {
          fmid = env->GetMethodID(clazz, "getFromRevision",
                                  "()" JAVAHL_ARG("/types/Revision;"));
          if (JNIUtil::isJavaExceptionThrown())
            return;
        }

      jobject jstartRevision = env->CallObjectMethod(jrange, fmid);
      if (JNIUtil::isExceptionThrown())
        return;

      Revision startRevision(jstartRevision);
      if (JNIUtil::isExceptionThrown())
        return;

      *range_start = *startRevision.revision();
      if (JNIUtil::isExceptionThrown())
        return;
    }

  if (range_end)
    {
      static jmethodID tmid = 0;
      if (tmid == 0)
        {
          tmid = env->GetMethodID(clazz, "getToRevision",
                                  "()" JAVAHL_ARG("/types/Revision;"));
          if (JNIUtil::isJavaExceptionThrown())
            return;
        }

      jobject jendRevision = env->CallObjectMethod(jrange, tmid);
      if (JNIUtil::isExceptionThrown())
        return;

      Revision endRevision(jendRevision);
      if (JNIUtil::isExceptionThrown())
        return;

      *range_end = *endRevision.revision();
      if (JNIUtil::isExceptionThrown())
        return;
    }

  if (range_inheritable)
    {
      static jmethodID imid = 0;
      if (imid == 0 && range_inheritable)
        {
          imid = env->GetMethodID(clazz, "isInheritable", "()Z");
          if (JNIUtil::isJavaExceptionThrown())
            return;
        }

      jboolean inheritable = env->CallBooleanMethod(jrange, imid);
      if (JNIUtil::isExceptionThrown())
        return;
      *range_inheritable = inheritable;
    }
}
} // anonymous namespace

svn_merge_range_t* RevisionRange::toMergeRange(SVN::Pool &pool) const
{
  svn_opt_revision_t range_start, range_end;
  svn_boolean_t range_inheritable;
  get_range_info(m_range, &range_start, &range_end, &range_inheritable);
  if (JNIUtil::isExceptionThrown())
    return NULL;

  if (range_start.kind != svn_opt_revision_number
      || range_end.kind != svn_opt_revision_number)
    JNIUtil::raiseThrowable("java.lang.InvalidStateException",
                            "Revsision ranges must contain revision numbers");

  svn_merge_range_t* range =
    static_cast<svn_merge_range_t*>
      (apr_palloc(pool.getPool(), sizeof(*range)));

  range->start = range_start.value.number;
  range->end = range_end.value.number;
  range->inheritable = range_inheritable;
  return range;
}

svn_opt_revision_range_t *RevisionRange::toRange(SVN::Pool &pool) const
{
  svn_opt_revision_range_t *range =
    static_cast<svn_opt_revision_range_t *>
      (apr_palloc(pool.getPool(), sizeof(*range)));

  get_range_info(m_range, &range->start, &range->end, NULL);
  if (JNIUtil::isExceptionThrown())
    range = NULL;
  return range;
}

jobject
RevisionRange::makeJRevisionRange(svn_merge_range_t *range)
{
    JNIEnv *env = JNIUtil::getEnv();

    jclass rangeClazz = env->FindClass(JAVAHL_CLASS("/types/RevisionRange"));
    if (JNIUtil::isJavaExceptionThrown())
        return NULL;
    static jmethodID rangeCtor = 0;
    if (rangeCtor == 0)
    {
        rangeCtor = env->GetMethodID(rangeClazz, "<init>", "(JJZ)V");
        if (JNIUtil::isJavaExceptionThrown())
            return NULL;
    }
    jobject jrange = env->NewObject(rangeClazz, rangeCtor,
                                    jlong(range->start),
                                    jlong(range->end),
                                    jboolean(range->inheritable));
    if (JNIUtil::isJavaExceptionThrown())
        return NULL;

    return jrange;
}

File: subversion/bindings/javahl/src/org/apache/subversion/javahl/types/RevisionRange.java
/**
 * @copyright
 * ====================================================================
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 * ====================================================================
 * @endcopyright
 */

package org.apache.subversion.javahl.types;


/**
 * Object that describes a revision range
 */
public class RevisionRange implements Comparable<RevisionRange>, java.io.Serializable
{
    // Update the serialVersionUID when there is a incompatible change made to
    // this class.  See the java documentation for when a change is incompatible.
    // http://java.sun.com/javase/7/docs/platform/serialization/spec/version.html#6678
    private static final long serialVersionUID = 2L;

    private Revision from;
    private Revision to;
    private boolean inheritable;

    /**
     * Creates a new instance.  Called by native library.
     */
    protected RevisionRange(long from, long to, boolean inheritable)
    {
        this.from = Revision.getInstance(from);
        this.to = Revision.getInstance(to);
        this.inheritable = inheritable;
    }

    /** @since 1.9 */
    public RevisionRange(Revision from, Revision to, boolean inheritable)
    {
        this.from = from;
        this.to = to;
        this.inheritable = inheritable;
    }

    public RevisionRange(Revision from, Revision to)
    {
        this.from = from;
        this.to = to;
        this.inheritable = true;
    }

    /**
     * Accepts a string in one of these forms: n m-n Parses the results into a
     * from and to revision
     * @param revisionElement revision range or single revision
     */
    public RevisionRange(String revisionElement)
    {
        super();
        if (revisionElement == null)
        {
            return;
        }

        this.inheritable = !revisionElement.endsWith("*");
        if (!this.inheritable)
            revisionElement =
                revisionElement.substring(0, revisionElement.length() - 1);

        int hyphen = revisionElement.indexOf('-');
        if (hyphen > 0)
        {
            try
            {
                long fromRev = Long
                        .parseLong(revisionElement.substring(0, hyphen));
                long toRev = Long.parseLong(revisionElement
                        .substring(hyphen + 1));
                this.from = new Revision.Number(fromRev);
                this.to = new Revision.Number(toRev);
            }
            catch (NumberFormatException e)
            {
                return;
            }

        }
        else
        {
            try
            {
                long revNum = Long.parseLong(revisionElement.trim());
                if (revNum <= 0)
                    return;
                this.to = new Revision.Number(revNum);
                this.from = new Revision.Number(revNum - 1);
            }
            catch (NumberFormatException e)
            {
                return;
            }
        }
    }

    public Revision getFromRevision()
    {
        return from;
    }

    public Revision getToRevision()
    {
        return to;
    }

    public boolean isInheritable()
    {
        return inheritable;
    }

    public String toString()
    {
        if (from != null && to != null)
        {
            String rep;

            if (from.getKind() == Revision.Kind.number
                && to.getKind() == Revision.Kind.number
                && (((Revision.Number)from).getNumber() + 1
                    == ((Revision.Number)to).getNumber()))
                rep = to.toString();
            else if (from.equals(to)) // Such ranges should never happen
                rep = from.toString();
            else
                rep = from.toString() + '-' + to.toString();
            if (!inheritable)
                return rep + '*';
            return rep;
        }
        return super.toString();
    }

    public static Long getRevisionAsLong(Revision rev)
    {
        long val = 0;
        if (rev != null && rev instanceof Revision.Number)
        {
            val = ((Revision.Number) rev).getNumber();
        }
        return new Long(val);
    }

    public int hashCode()
    {
        final int prime = 31;
        int result = (inheritable ? 1 : 2);
        result = prime * result + ((from == null) ? 0 : from.hashCode());
        result = prime * result + ((to == null) ? 0 : to.hashCode());
        return result;
    }

    /**
     * @param range The RevisionRange to compare this object to.
     */
    public boolean equals(Object range)
    {
        if (this == range)
            return true;
        if (!super.equals(range))
            return false;
        if (getClass() != range.getClass())
            return false;

        final RevisionRange other = (RevisionRange) range;

        if (from == null)
        {
            if (other.from != null)
                return false;
        }
        else if (!from.equals(other.from))
        {
            return false;
        }

        if (to == null)
        {
            if (other.to != null)
                return false;
        }
        else if (!to.equals(other.to))
        {
            return false;
        }

        return (inheritable == other.inheritable);
    }

    /**
     * <b>Note:</b> Explicitly ignores inheritable state.
     *
     * @param range The RevisionRange to compare this object to.
     */
    public int compareTo(RevisionRange range)
    {
        if (this == range)
            return 0;

        Revision other = (range).getFromRevision();
        return RevisionRange.getRevisionAsLong(this.getFromRevision())
            .compareTo(RevisionRange.getRevisionAsLong(other));
    }
}
